tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,426 @@
+            unwrapped_tensor_args.append(
+                f"auto {unwrapped_name} = at::native::empty_strided_meta({a_}.sizes(), {a_}.strides(), \
+/*dtype=*/c10::make_optional({a_}.scalar_type()), /*layout=*/c10::make_optional({a_}.layout()), \
+/*device=*/c10::make_optional(c10::Device(kMeta)), /*pin_memory=*/c10::nullopt);"
+            )
+            context.append(arg.with_name(unwrapped_name))
+        else:
+            # for non-tensor inputs, we want to pass them directly into the redispatch calls.
+            context.append(arg)
+    unwrap_tensor_args_str = '\n      '.join(unwrapped_tensor_args)
+    return unwrap_tensor_args_str, context
+
+# The function lambdas generated for each view op in the functionalization pass are of the form
+# [captures](input_args) -> return_type {
+#     return name(call_args);
+# }
+# This function returns the tuple of (input_args, name, return_type, call_args)
+# One thing to note: for multi-output views, the forward lambda needs to call the op and index into the output.
+# i.e. `return op(call_args)[idx]`
+# In this case, the <call_args> binding list returned here includes the index argument,
+# and the caller is responsible for parsing it out when generating the call-site.
+def view_lambda_types(f: NativeFunction, *, reverse: bool) -> Tuple[List[Binding], str, CType, List[Binding]]:
+    base_binding = Binding(
+        name='base',
+        nctype=NamedCType(name='base', type=ConstRefCType(BaseCType(tensorT))),
+        argument=Argument(name='base', type=BaseType(BaseTy.Tensor), default=None, annotation=None),
+        default=None)
+    mutated_view_binding = Binding(
+        name='mutated_view',
+        nctype=NamedCType(name='mutated_view', type=ConstRefCType(BaseCType(tensorT))),
+        argument=Argument(name='base', type=BaseType(BaseTy.Tensor), default=None, annotation=None),
+        default=None)
+    mutated_view_idx_binding = Binding(
+        name='mutated_view_idx',
+        nctype=NamedCType(name='mutated_view_idx', type=BaseCType(intT)),
+        argument=Argument(name='base', type=BaseType(BaseTy.Tensor), default=None, annotation=None),
+        default=None)
+
+    non_self_args = dispatcher.jit_arguments(f.func, skip_self=True)
+    for a in non_self_args:
+        # There currently aren't any views that take extra tensor args aside from self.
+        # If we eventually have any, the codegen will need extra logic to support it - for example,
+        # when we call the at::meta:: API, all tensor args need to be converted to the meta device.
+        assert not a.type.is_tensor_like()
+    non_self_bindings = [dispatcher.argument(a) for a in non_self_args]
+
+    # View ops always return a tensor or list of tensors.
+    assert len(f.func.returns) > 0
+    assert f.func.returns[0].type == BaseType(BaseTy.Tensor) or \
+        (isinstance(f.func.returns[0].type, ListType) and f.func.returns[0].type.elem == BaseType(BaseTy.Tensor))
```

Comment:
If len(f.func.returns) > 1, shouldn't you also be asserting on the higher indexed returns?